### PR TITLE
[ISSUE #210]Support 4.4.0 ACL

### PR DESCRIPTION
**Pull request by stateIs0**
## What is the purpose of the change

Support RocketMQ broker 4.4.0 ACL feature.

## Brief changelog

Support RocketMQ broker 4.4.0 ACL feature.

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by zongtanghu** [comment]
Hi @stateIs0  Your pr's code has some conflicts.Please resove it firstly.

**Comment by walking98** [comment]
it will be better if we can set different ACL AK/SK for each Topic in front page.

**Comment by zongtanghu** [comment]
> it will be better if we can set different ACL AK/SK for each Topic in front page.

Yes,but the goal of this pr just makes console project to support ACL.So we can put forward another issue for set different ACL AK/SK for each Topic in front page.

**Comment by walking98** [comment]
Hi @stateIs0 
The fix will make mqadmin invocation pass when config ACL ak/sk of admin role in the spring configuration file.

I fell it still need add the ACL RPCHook in other two places for sending and consuming messages:
···
1. org.apache.rocketmq.console.service.impl.TopicServiceImpl     
  DefaultMQProducer producer = new DefaultMQProducer(MixAll.SELF_TEST_PRODUCER_GROUP);

2.  org.apache.rocketmq.console.service.impl.MessageServiceImpl
    DefaultMQPullConsumer consumer = new DefaultMQPullConsumer(MixAll.TOOLS_CONSUMER_GROUP, null);
···

**Comment by stateIs0** [comment]
@walking98 ok,wait a moment

**Comment by walking98** [comment]
> @walking98 ok,wait a moment

@stateIs0  Do you have any plan for this PR? it needs some polish.

**Comment by stateIs0** [comment]
@walking98 i'm so Sorry, my front-end technology is not good, I am modifying the front-end code, and the new pr will be commited this weekend, is that ok?

File: rocketmq-console/src/main/java/org/apache/rocketmq/console/support/AclClientRPCHookFactory.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.apache.rocketmq.acl.common.AclClientRPCHook;
+import org.apache.rocketmq.acl.common.SessionCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Acl Helper.
+ *
+ * @author 莫那·鲁道
```
Comment (walking98) [inline rocketmq-console/src/main/java/org/apache/rocketmq/console/support/AclClientRPCHookFactory.java]: It is not suggested to add auth in Apache way

File: rocketmq-console/src/main/resources/application.properties [hunk 2]
```diff
@@ -14,5 +14,9 @@ rocketmq.config.isVIPChannel=
 rocketmq.config.dataPath=/tmp/rocketmq-console/data
 #set it false if you don't want use dashboard.default true
 rocketmq.config.enableDashBoardCollect=true
+# rocketMQ server default admin account
+rocketmq.config.accessKey=rocketmq2
```
Comment (walking98) [inline rocketmq-console/src/main/resources/application.properties]: not to set the ak/sk properties by default for backward compatible, namely, disable ACL by default

Comment (stateIs0) [reply rocketmq-console/src/main/resources/application.properties]: If ak or sk is null, the acl configuration will not take effect and the acl functionality will be disabled.

**Comment by walking98** [comment]
LGTM

**Comment by stateIs0** [comment]
thx

**Comment by walking98** [comment]
@stateIs0 please try to resolve the conflict according to this link:http://rocketmq.apache.org/docs/pull-request/, or the PR can not be merged later on.

@zongtanghu @wlliqipeng  Need you guys review and merge

**Comment by stateIs0** [comment]
@walking98 Is there anything else I can do for you?

File: rocketmq-console/src/main/java/org/apache/rocketmq/console/support/AclClientRPCHookFactory.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    private static class AclClientRPCHookFactoryLazyHolder {
+
+        private static final AclClientRPCHookFactory INSTANCE = new AclClientRPCHookFactory();
+    }
+
+
+    public AclClientRPCHook createAclClientRPCHook(String accessKey, String secretKey) {
+
+        if (accessKey != null && secretKey != null) {
```
Comment (zongtanghu) [inline rocketmq-console/src/main/java/org/apache/rocketmq/console/support/AclClientRPCHookFactory.java]: Here,I think it's nonvaluable if the accessKey or secretKey is empty String.So you can judge whether they is empty or not?

File: rocketmq-console/src/main/java/org/apache/rocketmq/console/aspect/admin/MQAdminAspect.java [hunk 4]
```diff
@@ -54,11 +61,12 @@ public Object aroundMQAdminMethod(ProceedingJoinPoint joinPoint) throws Throwabl
             MethodSignature signature = (MethodSignature)joinPoint.getSignature();
             Method method = signature.getMethod();
             MultiMQAdminCmdMethod multiMQAdminCmdMethod = method.getAnnotation(MultiMQAdminCmdMethod.class);
+            String accessKey = rmqConfigure.getAccessKey();
```
Comment (zongtanghu) [inline rocketmq-console/src/main/java/org/apache/rocketmq/console/aspect/admin/MQAdminAspect.java]: Maybe,users can set null or emtpy String from config file.So,you should judge whether they is empty or not?If users open the aclenable flag and the accessKey or secretKey value is null,you can print the warning or erroring informations in the log.And,here you must think of another condition that users may not want to use the acl feature!

**Comment by zongtanghu** [comment]
Otherwise,the pr should need some unit tests! @stateIs0 @walking98

File: rocketmq-console/src/main/resources/static/src/topic.js [hunk 5]
```diff
@@ -363,6 +363,13 @@ module.controller('sendTopicMessageDialogController', ['$scope', 'ngDialog', '$h
             topic: $scope.ngDialogData.topic,
             key: "key",
             tag:"tag",
+<<<<<<< HEAD
```
Comment (ox123) [inline rocketmq-console/src/main/resources/static/src/topic.js]: Conflict was not resolved,should resolve it before merge into branch

**Comment by walking98** [comment]
It will be better to add a README_ACL.md to tell how to set the acl properties and steps about configuration/usages of the ACL in the console. @stateIs0 

@zongtanghu Do you think okay to merge this PR first and polish it later with the full ACL setting feature together?

**Comment by stateIs0** [comment]
@walking98 What do you think if I write the use guide for acl feature in rocketmq-console/doc/1_0_0/ userguide_cn.md?

**Comment by walking98** [comment]
> @walking98 What do you think if I write the use guide for acl feature in rocketmq-console/doc/1_0_0/ userguide_cn.md?

I think fine if you add extra section for the ACL usage; or add a new file beneath the directory. @stateIs0

**Comment by zongtanghu** [comment]
@stateIs0 Hi can I join you wechat or dingding talk?We can discuss this pr and help you to finish it.

**Comment by stateIs0** [comment]
ok ，I already in dingding talk.😄

Hu Zongtang <notifications@github.com>于2019年7月15日 周一下午10:13写道：

**Comment by zongtanghu** [comment]
What's you dingding talk id,I can join you!

File: rocketmq-console/pom.xml [hunk 6]
```diff
@@ -89,6 +89,18 @@
...
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-beans</artifactId>
+            <version>4.3.20.RELEASE</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-beans</artifactId>
```
Comment (odbozhou) [inline rocketmq-console/pom.xml]: Repeated dependency

**Comment by honglouleiyan** [comment]
> 你在说什么，我可以加入你们！
What's you dingding talk id,I also want join you! So as to know the on-line time first，I want to integrate secondary development into our system！

**Comment by vongosling** [comment]
Any update for this pr ?

**Comment by syncwt** [comment]
I have added AK/SK config in resources/application.properties. But it doesn't work at all.Is the current master branch fix the issue?
The Error info as list
`java.lang.RuntimeException: org.apache.rocketmq.client.exception.MQBrokerException: CODE: 1  DESC: org.apache.rocketmq.acl.common.AclException: No accessKey is configured, org.apache.rocketmq.acl.plain.PlainPermissionManager.validate(PlainPermissionManager.java:394)
For more information, please visit the url, http://rocketmq.apache.org/docs/faq/
	at com.google.common.base.Throwables.propagate(Throwables.java:160)
	at org.apache.rocketmq.console.task.DashboardCollectTask.collectTopic(DashboardCollectTask.java:157)
	at org.apache.rocketmq.console.task.DashboardCollectTask$$FastClassBySpringCGLIB$$c818b8cd.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:88)
	at org.apache.rocketmq.console.aspect.admin.MQAdminAspect.aroundMQAdminMethod(MQAdminAspect.java:63)
	at sun.reflect.GeneratedMethodAccessor43.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:644)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:633)
	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:70)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
	at org.apache.rocketmq.console.task.DashboardCollectTask$$EnhancerBySpringCGLIB$$20cf5d8b.collectTopic(<generated>)
	at sun.reflect.GeneratedMethodAccessor49.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:93)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.rocketmq.client.exception.MQBrokerException: CODE: 1  DESC: org.apache.rocketmq.acl.common.AclException: No accessKey is configured, org.apache.rocketmq.acl.plain.PlainPermissionManager.validate(PlainPermissionManager.java:394)
For more information, please visit the url, http://rocketmq.apache.org/docs/faq/
	at org.apache.rocketmq.client.impl.MQClientAPIImpl.queryTopicConsumeByWho(MQClientAPIImpl.java:1673)
	at org.apache.rocketmq.tools.admin.DefaultMQAdminExtImpl.queryTopicConsumeByWho(DefaultMQAdminExtImpl.java:631)
	at org.apache.rocketmq.tools.admin.DefaultMQAdminExt.queryTopicConsumeByWho(DefaultMQAdminExt.java:381)
	at org.apache.rocketmq.console.service.client.MQAdminExtImpl.queryTopicConsumeByWho(MQAdminExtImpl.java:320)
	at org.apache.rocketmq.console.service.client.MQAdminExtImpl$$FastClassBySpringCGLIB$$f6279829.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:88)
	at org.apache.rocketmq.console.aspect.admin.MQAdminAspect.aroundMQAdminMethod(MQAdminAspect.java:63)
	at sun.reflect.GeneratedMethodAccessor43.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:644)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:633)
	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:70)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
	at org.apache.rocketmq.console.service.client.MQAdminExtImpl$$EnhancerBySpringCGLIB$$6ba4751f.queryTopicConsumeByWho(<generated>)
	at org.apache.rocketmq.console.task.DashboardCollectTask.collectTopic(DashboardCollectTask.java:88)
	... 33 common frames omitted`

**Pull request closed without merging**